automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend

```diff
@@ -646,220 +696,54 @@ namespace canmapping {
 */
 	::automotive::GenericCANMessage «className»::encode(odcore::data::Container &c) 
 	{
-    «IF canIDs.size>0»
-
-		«IF canIDs.size>1»
-			cerr<<"Warning: Mapping '«className»' is defined over more than one concrete CAN messages. Only one of those CAN messages will be returned."<<endl;
-	    «ENDIF»
-
-		«var String gcmPrefix="GCM_"»
-		«var String gcmPayloadPrefix="GCMPayload_"»
-		«FOR id:canIDs»
-			::automotive::GenericCANMessage «gcmPrefix+id»;
-			«gcmPrefix+id».setIdentifier(«id»);
-			uint64_t «gcmPayloadPrefix+id»=0x0;
-   		«ENDFOR»
-		
-			if(c.getDataType() != ::«mapping.mappingName.replaceAll("\\.","::")»::ID())
-			{
-				// something went wrong
-				::automotive::GenericCANMessage gcm;
-				gcm.setData(0x0);
-				gcm.setLength(0);
-				return gcm;
-			}
+	«IF canIDs.size==0»
+	(void)c;
+	// Return an empty GenericCANMessage
+	cerr<<"Fatal Error: Mapping '«className»' is empty."<<endl;
+	::automotive::odcantools::CANMessage error(0x00,0,0x00);
+	return error.getGenericCANMessage();
+    «ELSEIF canIDs.size<0»
+	(void)c;
+	cerr<<"Fatal Error: Mapping '«className»' experienced an error. "<<endl;
+	::automotive::odcantools::CANMessage error(0x00,0,0x00);
+	return error.getGenericCANMessage();
+    «ELSE»
+    «{//make sure all the CAN messages exist*/
+        
+    }»
 
-			bool found, extracted, abort=false;		
-			::«mapping.mappingName.replaceAll("\\.","::")» temp«className»=c.getData<::«mapping.mappingName.replaceAll("\\.","::")»>();
-			odcore::reflection::MessageFromVisitableVisitor mfvv;
-			temp«className».accept(mfvv);
-			odcore::reflection::Message msg=mfvv.getMessage();
+		«var String cmNamePrefix="CM_" /* multiple canIDs partially supported */»
```

Comment:
It is related to https://github.com/se-research/OpenDaVINCI/blob/2017Q1.refactoring.cangenerator/automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend#L763 , so if a mapping is defined over a set of several physical CAN IDs, the encode will return only the first CAN message of the set (the decode still works as expected, though).